pysmt/rewritings.py

```diff
@@ -673,6 +679,126 @@ def walk_minus(self, formula, args, **kwargs):
 # EOC TimesDistributivity
 
 
+
+class Ackermannization():
+    def __init__(self, environment=None):
+        #funs_to_args keeps for every function symbol f,
+        #a set of lists of arguments.
+        #if f(g(x),y) and f(x,g(y)) occur in a formula, then we
+        #will have "f": set([g(x), y], [x, g(y)])
+        self._funs_to_args = {}
+
+        #maps the actual applications to the constants that will be
+        #generated
+        self._terms_to_consts = {}
+
+        #maps each application to an index, used for the constant generation
+        self._indexes = {}
+
+
+    def do_ackermannization(self, formula):
+        self._fill_maps(formula)
+        #new formula without function symbols
+        substitued_formula = self._make_substitutions(formula)
+        #function consistency
+        implications = self._get_equality_implications()
+        function_consistency = And(implications)
+        if (len(implications) == 0):
+            result = substitued_formula
+        else:
+            result = And(function_consistency, substitued_formula)
+        return result
+
+    def _get_equality_implications(self):
+        result = set([])
+        for f in self._funs_to_args:
+            implications = self._generate_implications(f)
+            result.update(implications)
+        return result
+
+
+    def _generate_implications(self, f):
+        result = set([])
+        possible_args = self._funs_to_args[f]
+        for option1, option2 in pairwise(possible_args):
+            implication = self._generate_implication(option1, option2, f)
+            result.add(implication)
+        return result
+
+    def _generate_implication(self, option1, option2, f):
+        left_conjuncts = set([])
+        for i in range(0, len(option1)):
+            const1 = self._terms_to_consts[option1[i]]
+            const2 = self._terms_to_consts[option2[i]]
+            conjunct = Equals(const1, const2)
```

Comment:
In case of UFs having a Boolean argument it would be better to use `EqualsOrIff` to avoid typing errors